File: C++/heapSort.cpp
#include<iostream>

using namespace std;

void display(float arr[],int size)
{
    for(int i=0;i<size;i++)
    {
        cout<<arr[i]<<"  ";
    }
    cout<<endl;
}


void heapify(float arr[],int size, int pos)
{
    int largest=pos;
    int left=2*pos+1;
    int right=2*pos+2;

    if(left< size && arr[left]>arr[largest])
    {
            largest=left;
    }
    if(right<size && arr[right]>arr[largest])
    {
            largest=right;
    }

    if(largest!=pos)
    {
        int temp = arr[pos];
        arr[pos]=arr[largest];
        arr[largest]=temp;

        display(arr,size);
        heapify(arr,size,largest);
    }
}

void heapsort(float arr[], int size)
{
    for(int i=size/2-1;i>=0;i--)
    {
        heapify(arr,size,i);
    }
    for(int i=size-1;i>=1;i--)
    {
        int temp=arr[0];
        arr[0]=arr[i];
        arr[i]=arr[0];

        heapify(arr,i,0);
    }
}


int main()
{
    int size;
    cout<<"\nEnter size: ";
   // cin>>size;

    /*float arr[size];
    cout<<"\nEnter elements: ";
    for(int i=0;i<size;i++)
        cin>>arr[i];
*/
    float arr[]={27,41,31,9,8};
    size= sizeof(arr)/sizeof(arr[0]);
    cout<<"\nEntered Array: ";
    display(arr,size);

    heapsort(arr,size);

    cout<<"\nSorted Array: ";
    display(arr,size);

    return 0;

}

File: Java/HeapSort.java
// Heap Sort in Java
  
  public class HeapSort {
  
    public void sort(int arr[]) {
      int len = arr.length;
      
     for (int i = len / 2 - 1; i >= 0; i--) {
        heapify(arr, len, i);
      }
  
      for (int i = len - 1; i >= 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
  
        heapify(arr, i, 0);
      }
    }
  
    void heapify(int arr[], int len, int i) {
      int largest = i;
      int l = 2 * i + 1;
      int r = 2 * i + 2;
  
      if (l < len && arr[l] > arr[largest])
        largest = l;
  
      if (r < len && arr[r] > arr[largest])
        largest = r;
  
      if (largest != i) {
        int swap = arr[i];
        arr[i] = arr[largest];
        arr[largest] = swap;
  
        heapify(arr, len, largest);
      }
    }
  
    static void printArray(int arr[]) {
      int length = arr.length;
      for (int i = 0; i < length; ++i)
        System.out.print(arr[i] + " ");
        System.out.println();
    }
  
    public static void main(String args[]) {
      int arr[] = { 3, 15, 10, 8, 5, 11 };
  
      HeapSort heapsort = new HeapSort();
      heapsort.sort(arr);
  
      System.out.println("Array after sorting: ");
      printArray(arr);
    }
  }
